docs/manual.md

```diff
@@ -807,91 +805,294 @@ user:
+      - 'group1'
+      - 'group2'
+      - 'group2' # this 2nd group2 is not in the returned content
+```
+
+#### io.Readers
+
+This is the most magical matcher for goss. It remains a default for historic and performance reasons. Some attributes return an io.Reader that is read line by line (ex. file content, command, http body). This allows goss to validate large files/content efficiently.
+
 
-* `"string"` - checks if any line contain string.
-* `"!string"` - inverse of above, checks that no line contains string
-* `"\\!string"` - escape sequence, check if any line contains `"!string"`
-* `"/regex/"` - verifies that line contains regex
-* `"!/regex/"` - inverse of above, checks that no line contains regex
+Each pattern is checked against the attribute output, the type of patterns are:
 
-**NOTE:** Pattern attributes do not support [Advanced Matchers](#advanced-matchers)
+* `"foo"` - checks if any line contains `foo`
+* `"!foo"` - inverse of above, checks that no line contains `foo`
+* `"\\!foo"` - escape sequence, check if any line contains `!string`
+* `"/[Rr]egex/"` - verifies that line matches regex
+* `"!/[Rr]egex/"` - inverse of above, checks that no line matches regex
 
-**NOTE:** Regex support is based on golang's regex engine documented [here](https://golang.org/pkg/regexp/syntax/)
+**NOTE:** Regex support is based on Golang's regex engine documented [here](https://golang.org/pkg/regexp/syntax/)
 
 **NOTE:** You will **need** the double backslash (`\\`) escape for Regex special entities, for example `\\s` for blank spaces.
 
-### Example
-```bash
-$ cat /tmp/test.txt
-found
-!alsofound
+Example:
+
+```yaml
+file:
+  /tmp/test.txt:
+    exists: true
+    contains:
+    - "foo"
+    - "!bar"
+    - "/[Gg]oss/"
+```
 
+The above is roughly similar to:
 
-$ cat goss.yaml
+```yaml
 file:
   /tmp/test.txt:
     exists: true
     contains:
-    - found
-    - /fou.d/
-    - "\\!alsofound"
-    - "!missing"
-    - "!/mis.ing/"
+      and:
+        - contain-element: "foo"
+        - not: {contain-element: "bar"}
+        - contain-element: {match-regexp: "[Gg]oss"}
+
+```
+
+### Transforms
+
+If the system state type and the expected type don't match, goss will attempt to transform the system state type before matching it.
+
+For example, kernel-param attribute returns a string, however, it can be tested using numeric comparisons:
+
+Example kernel-param test:
+```yaml
+kernel-param:
+  net.core.somaxconn:
+      value: "128"
+```
+
+Kernel-param (failing) test with transform:
+```yaml
+kernel-param:
+  net.core.somaxconn:
+      value: {gt: 200}
+```
+
+When a transformed test fails, it will detail the transformers used, the `-o include_raw` option can be used to include the raw, untransformed attribute value:
+```
+$ goss v
+F
+
+Failures/Skipped:
+
+KernelParam: net.core.somaxconn: value:
+Expected
+    128
+to be >
+    200
+the transform chain was
+    [{"to-numeric":{}}]
+
+Total Duration: 0.001s
+Count: 1, Failed: 1, Skipped: 0
```

Comment:
As a user, I would find this easier to interpret if
* the front of the messaging included the input value and type
* the end of the messaging showed the desired output type

So, sketching something like this?

```suggestion
KernelParam: net.core.somaxconn: value:
Expected
    (string)  128
to be >
    (numeric) 200
the transform chain was
    [{"to-numeric":{ input: "128" }}]

Total Duration: 0.001s
Count: 1, Failed: 0, Skipped: 0, Error: 1
```

I put the types ahead of the values since the width of the type-string-names is predictable, and then the lines can be padded out to the longest one so that the values are all going to appear starting from the same column, making them easier to eyeball.

I added `Error: 1` because it feels to me that a failure to transform is an internal error as opposed to a failed assertion. That struck me as something I'd like to know separately, since it means there's something wrong in how I've defined my assertions, rather than a for-sure assertion-failure.

Reply:
This suggestion accidentally also (I think?) removes the need for the `-o include_raw`, since it includes the same information with no extra output lines needed.

Reply:
Gonna defer this conversation to regular github comments to make it easier to track. I'll detail the different use-cases there.